mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp

```diff
@@ -114,15 +145,63 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
   results.add<SqrtReciprocalOptimization>(context);
 }
 
-LogicalResult SelectOp::canonicalize(SelectOp op, PatternRewriter &rewriter) {
-  auto notOp = op.getPred().getDefiningOp<tosa::LogicalNotOp>();
-  if (!notOp)
-    return failure();
-  rewriter.modifyOpInPlace(op, [&]() {
-    op.getOperation()->setOperands(
-        {notOp.getInput1(), op.getOnFalse(), op.getOnTrue()});
-  });
-  return success();
+struct SelectLogicalNotOptimization : public OpRewritePattern<tosa::SelectOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(tosa::SelectOp op,
+                                PatternRewriter &rewriter) const override {
+    auto notOp = op.getPred().getDefiningOp<tosa::LogicalNotOp>();
+    if (!notOp)
+      return failure();
+    rewriter.modifyOpInPlace(op, [&]() {
+      op.getOperation()->setOperands(
+          {notOp.getInput1(), op.getOnFalse(), op.getOnTrue()});
+    });
+    return success();
+  }
+};
+
+// This pattern canonicalizaties the following pattern:
+// %0 = tosa.greater_equal(input, 0)
+// %1 = tosa.select(%0, input, 0)
+// to tosa.clamp{min = 0, max = max}(input)
+// This pattern occurs in decompositions of LeakyReLU/PReLU with an alpha
+// of zero
+struct SelectToClampOptimization : public OpRewritePattern<tosa::SelectOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(tosa::SelectOp op,
+                                PatternRewriter &rewriter) const override {
+    auto onFalse = op.getOnFalse();
+    DenseElementsAttr onFalseAttr;
+    if (!matchPattern(onFalse, m_Constant(&onFalseAttr)) ||
+        !isSplatZero(onFalse.getType().getElementType(), onFalseAttr)) {
+      return failure();
```

Comment:
can we change these to return rewrite.notifymatchfailure with a message.

Reply:
Done